src/deploy/functions/deploymentPlanner.ts

```diff
@@ -0,0 +1,175 @@
+  entryPoint: string;
+  runtime?: string;
+  vpcConnector?: string;
+  vpcConnectorEgressSettings?: string;
+  ingressSettings?: string;
+  availableMemoryMb?: number;
+  timeout?: number;
+  maxInstances?: number;
+  serviceAccountEmail?: string;
+  httpsTrigger?: any;
+  eventTrigger?: any;
+  failurePolicy?: {};
+  schedule?: object;
+  timeZone?: string;
+  regions?: string[];
+}
+
+export interface RegionMap {
+  [region: string]: CloudFunctionTrigger[];
+}
+
+export interface RegionalDeployment {
+  region: string;
+  sourceToken?: string;
+  firstFunctionDeployment?: CloudFunctionTrigger;
+  functionsToCreate: CloudFunctionTrigger[];
+  functionsToUpdate: CloudFunctionTrigger[];
+  schedulesToCreateOrUpdate: CloudFunctionTrigger[];
+}
+
+export interface DeploymentPlan {
+  regionalDeployments: RegionalDeployment[];
+  functionsToDelete: string[];
+  schedulesToDelete: string[];
+}
+
+/**
+ * Creates a map of regions to all the CloudFunctions being deployed
+ * to that region.
+ * @param projectId The project in use.
+ * @param parsedTriggers A list of all CloudFunctions in the deployment.
+ */
+export function createFunctionsByRegionMap(
+  projectId: string,
+  parsedTriggers: CloudFunctionTrigger[]
+): RegionMap {
+  const regionMap: RegionMap = {};
+  for (const trigger of parsedTriggers) {
+    if (!trigger.regions) {
+      trigger.regions = ["us-central1"];
+    }
+    // Create a separate CloudFunction for
+    // each region we deploy a function to
+    for (const region of trigger.regions) {
+      const triggerDeepCopy = JSON.parse(JSON.stringify(trigger));
+      if (triggerDeepCopy.regions) {
+        delete triggerDeepCopy.regions;
+      }
+      triggerDeepCopy.name = [
+        "projects",
+        projectId,
+        "locations",
+        region,
+        "functions",
+        trigger.name,
+      ].join("/");
+      regionMap[region] = regionMap[region] || [];
+      regionMap[region].push(triggerDeepCopy);
+    }
+  }
+  return regionMap;
+}
+
+/**
+ * Helper method to turn a RegionMap into a flat list of all functions in a deployment.
+ * @param regionMap A RegionMap for the deployment.
+ */
+export function flattenRegionMap(regionMap: RegionMap): CloudFunctionTrigger[] {
+  const triggers: CloudFunctionTrigger[] = [];
+  for (const [k, v] of Object.entries(regionMap)) {
+    triggers.push(...v);
+  }
+  return triggers;
+}
+
+/**
+ * Create a plan for deploying all functions in one region.
+ * @param region The region of this deployment
+ * @param functionsInSourceByRegion The functions present in the code currently being deployed.
+ * @param existingFunctionNames The names of all functions that already exist.
+ * @param existingScheduledFunctionNames The names of all schedules functions that already exist.
+ * @param filters The filters, passed in by the user via  `--only functions:`
+ */
+export function createDeploymentPlan(
+  functionsInSourceByRegion: RegionMap,
```

Comment:
Is there a reason to have this take a `RegionMap` instead of having the region grouped functions be an implementation detail of creating a deployment plan? Should the signature be instead

```ts
export function createDeploymentPlan(
  projectId: string,
  parsedTriggers: CloudFunctionTrigger[],
  existingFunctions: CloudFunctionTrigger[],
  filters: string[][]
): DeploymentPlan {
```